File: incre-tests/synduce/tailopt/sum.f
Inductive List = nil Unit | cons {Int, List};

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

snoc = fix (
  \f: List -> Int -> List. \xs: List. \w: Int.
  match xs with
    nil _ -> cons {w, nil unit}
  | cons {h, t} -> cons {h, f t w}
  end
);

repr = (fix (
  \f: Compress List -> List -> Compress List.
  \pre: Compress List. \xs: List.
  match xs with
    nil _ -> pre
  | cons {h, t} -> f (snoc pre h) t
  end
)) (nil unit);

main = \xs: List. spec (repr xs);

File: ml-benchmark/synduce/tailopt/sum.ml
type l = Nil of unit | Cons of int * l

let rec spec xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) -> h + spec t

let rec snoc xs w =
  match xs with
  | Nil _ -> Cons (w, Nil ())
  | Cons (h, t) -> Cons (h, snoc t w)

let rec repr_aux pre xs =
  match xs with
  | Nil _ -> pre
  | Cons (h, t) -> repr_aux (snoc pre h) t

let repr xs = repr_aux (Nil ()) xs

let program xs = spec (repr xs)